ngraph/core/reference/include/ngraph/runtime/reference/convolution.hpp

```diff
@@ -24,338 +25,219 @@
-                    CoordinateTransform::Iterator in_it_end = in_transform.end();
-                    CoordinateTransform::Iterator filter_it_end = filter_transform.end();
-
-                    size_t in_channel_stride = row_major_strides(in_shape).at(in_channel_axis);
-                    size_t filter_in_channel_stride =
-                        row_major_strides(filter_shape).at(filter_in_channel_axis);
+                    std::vector<int> strides;
+                    std::vector<int> dilation;
+                    std::vector<int> pads_begin;
+                    std::vector<int> pads_end;
+
+                    ConvolutionParams(const Strides& strides_,
+                                      const Strides& dilation_,
+                                      const CoordinateDiff& pads_begin_,
+                                      const CoordinateDiff& pads_end_)
+                        : strides{strides_.begin(), strides_.end()}
+                        , dilation{dilation_.begin(), dilation_.end()}
+                        , pads_begin{pads_begin_.begin(), pads_begin_.end()}
+                        , pads_end{pads_end_.begin(), pads_end_.end()} {};
+                };
+
+                template <typename Int>
+                constexpr inline bool in_range(Int val, std::pair<Int, Int> range) noexcept
+                {
+                    return val >= range.first && val < range.second;
+                }
 
-                    while (in_it != in_it_end && filter_it != filter_it_end)
+                template <typename INPUT, typename FILTER, typename OUTPUT, typename ACCU>
+                void convolve_3D_channels(const ConvolutionParams& p,
+                                          const INPUT* batch,
+                                          const Shape& batch_shape,
+                                          const FILTER* filter,
+                                          const Shape& filter_shape,
+                                          OUTPUT*& out)
+                {
+                    const int input_size_z = batch_shape[1];
+                    const int input_size_y = batch_shape[2];
+                    const int input_size_x = batch_shape[3];
+                    const int filter_size_z = filter_shape[1];
+                    const int filter_size_y = filter_shape[2];
+                    const int filter_size_x = filter_shape[3];
+                    const int dilated_filter_size_z =
+                        filter_size_z + (filter_size_z - 1) * (p.dilation[0] - 1);
+                    const int dilated_filter_size_y =
+                        filter_size_y + (filter_size_y - 1) * (p.dilation[1] - 1);
+                    const int dilated_filter_size_x =
+                        filter_size_x + (filter_size_x - 1) * (p.dilation[2] - 1);
+
+                    const Shape input_channel_shape(++batch_shape.begin(), batch_shape.end());
```

Comment:
You may use `Span` from `reference/utils` but it probably doesn't bring any significant impact.

Reply:
@ilyachur what do you think about this? Is this syntactic sugar we need or not ? Asking because don't won't to rework it again :)

Reply:
If we don't have any improvements from `Span` usage, I am ok with the current approach.